File: obstacle_detector/obstacle_detector/obstacle_detector_node.py
#!/usr/bin/env python3
"""
Obstacle Detector Node
======================
Subscribes to a 2-D LiDAR scan, applies a configurable box filter, clusters
the filtered points, and publishes an obstacle-detected flag.

All tunable values live in  config/params.yaml — edit that file then relaunch.

Published topics
----------------
  /obstacle_detected   std_msgs/Bool   True  = obstacle present
                                       False = obstacle cleared

  /coverage_stop       std_msgs/Bool   mirrors /obstacle_detected when auto_pause=true
                                       (pauses / resumes coverage_path_follower)

Subscribed topics
-----------------
  /scan  (configurable)   sensor_msgs/LaserScan

Parameters — see config/params.yaml for descriptions
------------------------------------------------------
  scan_topic        str    default: /scan
  x_min / x_max     float  box filter X bounds  (default ±0.5 m)
  y_min / y_max     float  box filter Y bounds  (default ±0.5 m)
  cluster_distance  float  max gap inside a cluster  (default 0.15 m)
  min_cluster_size  int    min points per cluster    (default 5)
  min_clusters      int    clusters needed to trigger (default 1)
  auto_pause        bool   publish /coverage_stop     (default true)
"""

import math

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from std_msgs.msg import Bool


class ObstacleDetectorNode(Node):

    def __init__(self):
        super().__init__('obstacle_detector_node')

        # ── Declare parameters ─────────────────────────────────────────────────
        self.declare_parameter('scan_topic',       '/scan')
        self.declare_parameter('x_min',            -0.5)
        self.declare_parameter('x_max',             0.5)
        self.declare_parameter('y_min',            -0.5)
        self.declare_parameter('y_max',             0.5)
        self.declare_parameter('cluster_distance',  0.15)
        self.declare_parameter('min_cluster_size',  5)
        self.declare_parameter('min_clusters',      1)
        self.declare_parameter('auto_pause',        True)

        # ── Load parameters ────────────────────────────────────────────────────
        self._x_min          = self.get_parameter('x_min').value
        self._x_max          = self.get_parameter('x_max').value
        self._y_min          = self.get_parameter('y_min').value
        self._y_max          = self.get_parameter('y_max').value
        self._cluster_dist   = self.get_parameter('cluster_distance').value
        self._min_cluster_sz = self.get_parameter('min_cluster_size').value
        self._min_clusters   = self.get_parameter('min_clusters').value
        self._auto_pause     = self.get_parameter('auto_pause').value
        scan_topic           = self.get_parameter('scan_topic').value

        # ── State ──────────────────────────────────────────────────────────────
        self._obstacle_active = False

        # ── Publishers ─────────────────────────────────────────────────────────
        self._pub_obstacle = self.create_publisher(Bool, '/obstacle_detected', 10)
        self._pub_stop     = self.create_publisher(Bool, '/coverage_stop',     10)

        # ── Subscriber ─────────────────────────────────────────────────────────
        self.create_subscription(LaserScan, scan_topic, self._cb_scan, 10)

        self.get_logger().info(
            f'Obstacle detector ready\n'
            f'  scan:    {scan_topic}\n'
            f'  box:     x[{self._x_min}, {self._x_max}]  '
            f'y[{self._y_min}, {self._y_max}]  (metres)\n'
            f'  cluster: size>={self._min_cluster_sz}  '
            f'gap<={self._cluster_dist} m  '
            f'min_clusters={self._min_clusters}\n'
            f'  auto_pause: {self._auto_pause}'
        )

    # ── Scan callback ──────────────────────────────────────────────────────────

    def _cb_scan(self, msg: LaserScan):
        # 1. Convert polar → Cartesian, keep only points inside the box
        points = []
        angle = msg.angle_min
        for r in msg.ranges:
            if math.isfinite(r) and msg.range_min <= r <= msg.range_max:
                x = r * math.cos(angle)
                y = r * math.sin(angle)
                if self._x_min <= x <= self._x_max and self._y_min <= y <= self._y_max:
                    points.append((x, y))
            angle += msg.angle_increment

        # 2. Cluster consecutive nearby points
        clusters = self._find_clusters(points)

        # 3. Evaluate trigger
        obstacle = len(clusters) >= self._min_clusters

        # 4. Publish only on state change
        if obstacle != self._obstacle_active:
            self._obstacle_active = obstacle
            self._pub_obstacle.publish(Bool(data=obstacle))
            if self._auto_pause:
                self._pub_stop.publish(Bool(data=obstacle))
            if obstacle:
                self.get_logger().warn(
                    f'OBSTACLE DETECTED: {len(clusters)} cluster(s) with '
                    f'{sum(len(c) for c in clusters)} points in box'
                )
            else:
                self.get_logger().info('Obstacle CLEARED — coverage can resume')

    # ── Clustering ────────────────────────────────────────────────────────────

    def _find_clusters(self, points: list) -> list:
        """
        Group scan-ordered points into clusters.
        A new cluster starts whenever the Euclidean gap between consecutive
        points exceeds cluster_distance.  Returns only clusters that meet
        min_cluster_size.
        """
        if not points:
            return []

        clusters = []
        current  = [points[0]]

        for i in range(1, len(points)):
            dx = points[i][0] - points[i - 1][0]
            dy = points[i][1] - points[i - 1][1]
            if math.sqrt(dx * dx + dy * dy) <= self._cluster_dist:
                current.append(points[i])
            else:
                if len(current) >= self._min_cluster_sz:
                    clusters.append(current)
                current = [points[i]]

        if len(current) >= self._min_cluster_sz:
            clusters.append(current)

        return clusters


# ── Entry point ───────────────────────────────────────────────────────────────

def main(args=None):
    rclpy.init(args=args)
    node = ObstacleDetectorNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: obstacle_detector/src/obstacle_detector_node.cpp
#include <cmath>
#include <memory>
#include <string>
#include <vector>

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/laser_scan.hpp>
#include <std_msgs/msg/bool.hpp>
#include <visualization_msgs/msg/marker_array.hpp>

#include <pcl/filters/passthrough.h>
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>
#include <pcl/search/kdtree.h>
#include <pcl/segmentation/extract_clusters.h>

class ObstacleDetectorNode : public rclcpp::Node
{
public:
  ObstacleDetectorNode() : Node("obstacle_detector_node")
  {
    // Declare parameters
    declare_parameter("scan_topic",      std::string("/scan"));
    declare_parameter("x_min",           -0.5);
    declare_parameter("x_max",            0.5);
    declare_parameter("y_min",           -0.5);
    declare_parameter("y_max",            0.5);
    declare_parameter("cluster_distance", 0.15);
    declare_parameter("min_cluster_size", 5);
    declare_parameter("min_clusters",     1);
    declare_parameter("vicinity_debounce", 5);   // consecutive frames needed to trigger vicinity
    declare_parameter("vicinity_x_min",  -0.6);
    declare_parameter("vicinity_x_max",   0.6);
    declare_parameter("vicinity_y_min",  -0.6);
    declare_parameter("vicinity_y_max",   0.6);

    // Load parameters
    x_min_          = get_parameter("x_min").as_double();
    x_max_          = get_parameter("x_max").as_double();
    y_min_          = get_parameter("y_min").as_double();
    y_max_          = get_parameter("y_max").as_double();
    cluster_dist_   = get_parameter("cluster_distance").as_double();
    min_cluster_sz_ = get_parameter("min_cluster_size").as_int();
    min_clusters_      = get_parameter("min_clusters").as_int();
    vicinity_debounce_ = get_parameter("vicinity_debounce").as_int();
    vx_min_            = get_parameter("vicinity_x_min").as_double();
    vx_max_         = get_parameter("vicinity_x_max").as_double();
    vy_min_         = get_parameter("vicinity_y_min").as_double();
    vy_max_         = get_parameter("vicinity_y_max").as_double();
    auto scan_topic = get_parameter("scan_topic").as_string();

    // Publishers
    pub_obstacle_  = create_publisher<std_msgs::msg::Bool>("/obstacle_detected",   10);
    pub_vicinity_  = create_publisher<std_msgs::msg::Bool>("/obstacle_in_vicinity",10);
    pub_markers_   = create_publisher<visualization_msgs::msg::MarkerArray>(
      "/obstacle_detector/markers", 10);

    // Subscriber
    sub_scan_ = create_subscription<sensor_msgs::msg::LaserScan>(
      scan_topic, 10,
      std::bind(&ObstacleDetectorNode::cb_scan, this, std::placeholders::_1));

    RCLCPP_INFO(
      get_logger(),
      "Obstacle detector ready\n"
      "  scan:    %s\n"
      "  box:     x[%.2f, %.2f]  y[%.2f, %.2f]  (metres)\n"
      "  cluster: size>=%d  gap<=%.2f m  min_clusters=%d",
      scan_topic.c_str(), x_min_, x_max_, y_min_, y_max_,
      min_cluster_sz_, cluster_dist_, min_clusters_);
  }

private:
  void cb_scan(const sensor_msgs::msg::LaserScan::SharedPtr msg)
  {
    // 1. Convert polar → Cartesian into a PCL cloud
    auto cloud = std::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
    double angle = msg->angle_min;
    for (float r : msg->ranges) {
      if (std::isfinite(r) && r >= msg->range_min && r <= msg->range_max) {
        cloud->emplace_back(
          static_cast<float>(r * std::cos(angle)),
          static_cast<float>(r * std::sin(angle)),
          0.0f);
      }
      angle += msg->angle_increment;
    }

    // 2. Box filter — PassThrough on X then Y
    auto cloud_filtered = std::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
    pcl::PassThrough<pcl::PointXYZ> pass;
    pass.setInputCloud(cloud);
    pass.setFilterFieldName("x");
    pass.setFilterLimits(static_cast<float>(x_min_), static_cast<float>(x_max_));
    pass.filter(*cloud_filtered);

    pass.setInputCloud(cloud_filtered);
    pass.setFilterFieldName("y");
    pass.setFilterLimits(static_cast<float>(y_min_), static_cast<float>(y_max_));
    pass.filter(*cloud_filtered);

    // 3. Euclidean clustering
    int num_clusters = 0;
    std::vector<pcl::PointIndices> cluster_indices;
    if (!cloud_filtered->empty()) {
      auto tree = std::make_shared<pcl::search::KdTree<pcl::PointXYZ>>();
      tree->setInputCloud(cloud_filtered);

      pcl::EuclideanClusterExtraction<pcl::PointXYZ> ec;
      ec.setClusterTolerance(static_cast<float>(cluster_dist_));
      ec.setMinClusterSize(min_cluster_sz_);
      ec.setMaxClusterSize(static_cast<int>(cloud_filtered->size()));
      ec.setSearchMethod(tree);
      ec.setInputCloud(cloud_filtered);
      ec.extract(cluster_indices);

      num_clusters = static_cast<int>(cluster_indices.size());
    }

    // 4. Publish on state change only
    bool obstacle = num_clusters >= min_clusters_;
    if (obstacle != obstacle_active_) {
      obstacle_active_ = obstacle;
      std_msgs::msg::Bool out;
      out.data = obstacle;
      pub_obstacle_->publish(out);
      if (obstacle) {
        RCLCPP_WARN(get_logger(), "OBSTACLE DETECTED: %d cluster(s) in box", num_clusters);
      } else {
        RCLCPP_INFO(get_logger(), "Obstacle CLEARED — coverage can resume");
      }
    }

    // 5. Vicinity box — larger box, same clustering thresholds
    auto cloud_vicinity = std::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
    pass.setInputCloud(cloud);
    pass.setFilterFieldName("x");
    pass.setFilterLimits(static_cast<float>(vx_min_), static_cast<float>(vx_max_));
    pass.filter(*cloud_vicinity);

    pass.setInputCloud(cloud_vicinity);
    pass.setFilterFieldName("y");
    pass.setFilterLimits(static_cast<float>(vy_min_), static_cast<float>(vy_max_));
    pass.filter(*cloud_vicinity);

    int vicinity_clusters = 0;
    if (!cloud_vicinity->empty()) {
      auto tree_v = std::make_shared<pcl::search::KdTree<pcl::PointXYZ>>();
      tree_v->setInputCloud(cloud_vicinity);

      std::vector<pcl::PointIndices> vic_indices;
      pcl::EuclideanClusterExtraction<pcl::PointXYZ> ec_v;
      ec_v.setClusterTolerance(static_cast<float>(cluster_dist_));
      ec_v.setMinClusterSize(min_cluster_sz_);
      ec_v.setMaxClusterSize(static_cast<int>(cloud_vicinity->size()));
      ec_v.setSearchMethod(tree_v);
      ec_v.setInputCloud(cloud_vicinity);
      ec_v.extract(vic_indices);
      vicinity_clusters = static_cast<int>(vic_indices.size());
    }

    // Debounce: require vicinity_debounce_ consecutive positive frames before going high;
    // clear immediately on the first negative frame.
    bool raw_vicinity = vicinity_clusters >= min_clusters_;
    if (raw_vicinity) {
      vicinity_consec_ = std::min(vicinity_consec_ + 1, vicinity_debounce_);
    } else {
      vicinity_consec_ = 0;
    }
    bool in_vicinity = (vicinity_consec_ >= vicinity_debounce_);
    if (in_vicinity != vicinity_active_) {
      vicinity_active_ = in_vicinity;
      std_msgs::msg::Bool vic_out;
      vic_out.data = in_vicinity;
      pub_vicinity_->publish(vic_out);
      if (in_vicinity) {
        RCLCPP_INFO(get_logger(), "Object entered vicinity: %d cluster(s)", vicinity_clusters);
      } else {
        RCLCPP_INFO(get_logger(), "Vicinity cleared");
      }
    }

    // 6. Publish RViz2 markers
    publish_markers(msg->header, *cloud_filtered, *cloud_vicinity, cluster_indices);
  }

  // ── Marker helpers ────────────────────────────────────────────────────────

  // Returns a point for use in LINE_STRIP / POINTS markers
  static geometry_msgs::msg::Point make_point(double x, double y, double z = 0.0)
  {
    geometry_msgs::msg::Point p;
    p.x = x; p.y = y; p.z = z;
    return p;
  }

  void publish_markers(
    const std_msgs::msg::Header & header,
    const pcl::PointCloud<pcl::PointXYZ> & filtered,
    const pcl::PointCloud<pcl::PointXYZ> & vicinity,
    const std::vector<pcl::PointIndices> & cluster_indices)
  {
    visualization_msgs::msg::MarkerArray arr;

    // ── Marker 0: obstacle box outline (yellow LINE_STRIP) ──────────────────
    {
      visualization_msgs::msg::Marker box;
      box.header  = header;
      box.ns      = "obstacle_detector";
      box.id      = 0;
      box.type    = visualization_msgs::msg::Marker::LINE_STRIP;
      box.action  = visualization_msgs::msg::Marker::ADD;
      box.scale.x = 0.02;
      box.color.r = 1.0f;
      box.color.g = 1.0f;
      box.color.b = 0.0f;
      box.color.a = 0.8f;
      box.pose.orientation.w = 1.0;
      box.points = {
        make_point(x_min_, y_min_),
        make_point(x_max_, y_min_),
        make_point(x_max_, y_max_),
        make_point(x_min_, y_max_),
        make_point(x_min_, y_min_),
      };
      arr.markers.push_back(box);
    }

    // ── Marker 10: vicinity box outline (cyan LINE_STRIP, dashed via thin line) ──
    {
      visualization_msgs::msg::Marker vbox;
      vbox.header  = header;
      vbox.ns      = "obstacle_detector";
      vbox.id      = 10;
      vbox.type    = visualization_msgs::msg::Marker::LINE_STRIP;
      vbox.action  = visualization_msgs::msg::Marker::ADD;
      vbox.scale.x = 0.015;
      vbox.color.r = 0.0f;
      vbox.color.g = 0.9f;
      vbox.color.b = 0.9f;
      vbox.color.a = 0.6f;
      vbox.pose.orientation.w = 1.0;
      vbox.points = {
        make_point(vx_min_, vy_min_),
        make_point(vx_max_, vy_min_),
        make_point(vx_max_, vy_max_),
        make_point(vx_min_, vy_max_),
        make_point(vx_min_, vy_min_),
      };
      arr.markers.push_back(vbox);
    }

    // ── Marker 11: vicinity-only points (cyan POINTS — inside vicinity but outside obstacle box) ──
    {
      visualization_msgs::msg::Marker vpts;
      vpts.header  = header;
      vpts.ns      = "obstacle_detector";
      vpts.id      = 11;
      vpts.type    = visualization_msgs::msg::Marker::POINTS;
      vpts.action  = visualization_msgs::msg::Marker::ADD;
      vpts.scale.x = 0.04;
      vpts.scale.y = 0.04;
      vpts.color.r = 0.0f;
      vpts.color.g = 0.9f;
      vpts.color.b = 0.9f;
      vpts.color.a = 0.7f;
      vpts.pose.orientation.w = 1.0;
      for (const auto & p : vicinity) {
        // skip points that are already shown by the obstacle box markers
        if (p.x >= x_min_ && p.x <= x_max_ && p.y >= y_min_ && p.y <= y_max_) {
          continue;
        }
        vpts.points.push_back(make_point(p.x, p.y));
      }
      arr.markers.push_back(vpts);
    }

    // ── Marker 1: filtered points not in any cluster (grey POINTS) ──────────
    {
      // Build a set of indices that belong to a cluster
      std::vector<bool> in_cluster(filtered.size(), false);
      for (const auto & ci : cluster_indices) {
        for (int idx : ci.indices) {
          in_cluster[idx] = true;
        }
      }

      visualization_msgs::msg::Marker noise_pts;
      noise_pts.header  = header;
      noise_pts.ns      = "obstacle_detector";
      noise_pts.id      = 1;
      noise_pts.type    = visualization_msgs::msg::Marker::POINTS;
      noise_pts.action  = visualization_msgs::msg::Marker::ADD;
      noise_pts.scale.x = 0.04;
      noise_pts.scale.y = 0.04;
      noise_pts.color.r = 0.6f;
      noise_pts.color.g = 0.6f;
      noise_pts.color.b = 0.6f;
      noise_pts.color.a = 0.6f;
      noise_pts.pose.orientation.w = 1.0;

      for (std::size_t i = 0; i < filtered.size(); ++i) {
        if (!in_cluster[i]) {
          noise_pts.points.push_back(make_point(filtered[i].x, filtered[i].y));
        }
      }
      arr.markers.push_back(noise_pts);
    }

    // ── Markers 2+: one POINTS marker per cluster, colour-coded ─────────────
    // Palette: cycles through a few distinct colours
    static const std::array<std::array<float, 3>, 6> kPalette = {{
      {1.0f, 0.2f, 0.2f},   // red
      {0.2f, 0.8f, 0.2f},   // green
      {0.2f, 0.6f, 1.0f},   // blue
      {1.0f, 0.6f, 0.0f},   // orange
      {0.8f, 0.2f, 0.8f},   // magenta
      {0.0f, 0.9f, 0.9f},   // cyan
    }};

    int cluster_id = 0;
    for (const auto & ci : cluster_indices) {
      const auto & col = kPalette[cluster_id % kPalette.size()];

      visualization_msgs::msg::Marker cpt;
      cpt.header  = header;
      cpt.ns      = "obstacle_detector";
      cpt.id      = 2 + cluster_id;
      cpt.type    = visualization_msgs::msg::Marker::POINTS;
      cpt.action  = visualization_msgs::msg::Marker::ADD;
      cpt.scale.x = 0.06;
      cpt.scale.y = 0.06;
      cpt.color.r = col[0];
      cpt.color.g = col[1];
      cpt.color.b = col[2];
      cpt.color.a = 1.0f;
      cpt.pose.orientation.w = 1.0;

      for (int idx : ci.indices) {
        cpt.points.push_back(make_point(filtered[idx].x, filtered[idx].y));
      }
      arr.markers.push_back(cpt);
      ++cluster_id;
    }

    // Delete stale cluster markers from previous frames that had more clusters
    // (send DELETE action for any ids beyond what we just published)
    for (int id = 2 + cluster_id; id < 2 + prev_num_clusters_; ++id) {
      visualization_msgs::msg::Marker del;
      del.header = header;
      del.ns     = "obstacle_detector";
      del.id     = id;
      del.action = visualization_msgs::msg::Marker::DELETE;
      arr.markers.push_back(del);
    }
    prev_num_clusters_ = cluster_id;

    pub_markers_->publish(arr);
  }

  // Parameters — obstacle box
  double x_min_, x_max_, y_min_, y_max_;
  double cluster_dist_;
  int    min_cluster_sz_, min_clusters_;
  int    vicinity_debounce_;

  // Parameters — vicinity box
  double vx_min_, vx_max_, vy_min_, vy_max_;

  // State
  bool obstacle_active_{false};
  bool vicinity_active_{false};
  int  vicinity_consec_{0};
  int  prev_num_clusters_{0};

  // ROS interfaces
  rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr           pub_obstacle_;
  rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr           pub_vicinity_;
  rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr pub_markers_;
  rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr       sub_scan_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ObstacleDetectorNode>());
  rclcpp::shutdown();
  return 0;
}
